File: point.py
import actions
import entities
import worldmodel
import pygame
import math
import random
import image_store
import occ_grid

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def within_bounds(self,world):
        return (self.x >= 0 and self.x < world.num_cols and
                self.y >= 0 and self.y < world.num_rows)
    
    def is_occupied(self, world):
        return (self.within_bounds(world) and
                occ_grid.get_cell(world.occupancy, self) != None)
    
    def distance_sq(self, p2):
        return (self.x - p2.x)**2 + (self.y - p2.y)**2
    
    def adjacent(self,pt2):
        return ((self.x == pt2.x and abs(self.y - pt2.y) == 1) or
              (self.y == pt2.y and abs(self.x - pt2.x) == 1))
    def next_position(self, world, dest_pt):
        horiz = actions.sign(dest_pt.x - self.x)
        new_pt = Point(self.x + horiz, self.y)
      
        if horiz == 0 or new_pt.is_occupied(world):
            vert = actions.sign(dest_pt.y - self.y)
            new_pt = Point(self.x, self.y + vert)

            if vert == 0 or new_pt.is_occupied(world):
                new_pt = Point(self.x, self.y)

        return new_pt
     

File: java/Point.java
import java.lang.Math;

public class Point
{
	protected int x;
	protected int y;
	private Entity none;

	public Point(int x, int y)
	{
	    this.x = x;
	    this.y = y;
	}

	public int x()
	{
		return this.x;
	}

	public int y()
	{
		return this.y;
	}

	public boolean withinBounds(WorldModel world)
	{
	    return (this.x >= 0 && this.x < world.num_cols()) &&
            (this.y >= 0 && this.y < world.num_rows());
	}

	public boolean isOccupied(WorldModel world)
	{
		if (this.withinBounds(world))
		{
			if (!(world.occupancy().getCell(this) instanceof Entity))
			{
				return true;
			}
			return false;
		} 
		return false;

	}

	public int distanceSq(Point p2)
	{
		return (this.x - p2.x())*(this.x - p2.x()) + (this.y - p2.y())*(this.y - p2.y());
	}

	public boolean adjacent(Point pt2)
	{
		return (this.x == pt2.x() && Math.abs(this.y - pt2.y()) == 1) ||
		    (this.y == pt2.y() && Math.abs(this.x - pt2.x()) == 1);
	}

	public Point nextPosition(WorldModel world, Point destination)
	{
		Sign actions = new Sign();
      int horiz = actions.sign(destination.x() - this.x());
      Point newPt = new Point(this.x() + horiz, this.y);

      if(horiz == 0 || newPt.isOccupied(world))
      {
			int vert = actions.sign(destination.y() - this.x);
        	newPt = new Point(this.x(), this.y() + vert);
        	if (vert == 0 || newPt.isOccupied(world))
        	{
        		newPt = new Point(this.x(), this.y());
        	}
      }
		return newPt;
	}

	public Point findOpenAround(WorldModel world, int distance)
	{
		for(int dy = -distance; dy < (distance + 1); dy ++)
		{
			for(int dx = -distance; dx < (distance + 1); dx ++)
			{
				Point newPt = new Point(this.x + dx, this.y + dy);
				if (newPt.withinBounds(world) && !(newPt.isOccupied(world)))
				{
					return newPt;
				}
			}
		}
		return null;
	}
}
